src/bootstrap/src/core/build_steps/perf.rs

```diff
@@ -1,12 +1,115 @@
+use std::fmt::{Display, Formatter};
 use std::process::Command;
 
 use crate::core::build_steps::compile::{Std, Sysroot};
 use crate::core::build_steps::tool::RustcPerf;
 use crate::core::builder::Builder;
 use crate::core::config::DebuginfoLevel;
 
+/// Performs profiling or benchmarking with [`rustc-perf`](https://github.com/rust-lang/rustc-perf)
+/// using a locally built compiler.
+#[derive(Debug, Clone, clap::Parser)]
+pub struct PerfArgs {
+    #[clap(subcommand)]
+    cmd: PerfCommand,
+
+    #[clap(flatten)]
+    opts: SharedOpts,
+}
+
+impl Default for PerfArgs {
+    fn default() -> Self {
+        Self { cmd: PerfCommand::Eprintln, opts: SharedOpts::default() }
+    }
+}
+
+#[derive(Debug, Clone, clap::Parser)]
+enum PerfCommand {
+    /// Run `profile_local eprintln`.
+    /// This executes the compiler on the given benchmarks and stores its stderr output.
+    Eprintln,
+    /// Run `profile_local samply`
+    /// This executes the compiler on the given benchmarks and profiles it with `samply`.
+    /// You need to install `samply`, e.g. using `cargo install --locked samply`.
+    Samply,
```

Comment:
I just ran `x perf samply` and it turns out it needs root access (or I have to update `/proc/sys/kernel/perf_event_paranoid`). We probably need to extend this documentation here.

Reply:
Right, so this is a bit tricky. In general, many of the profilers require some custom configuration of the system, and also some postprocessing. This is documented in `rustc-perf`, and I'm not sure if it is a good idea to duplicate the documentation here.

The tool should let you know exactly what you should do w.r.t. `perf_event_paranoid`, so it is self-documenting in this sense.

Furthermore, even if you run the profiler, it just generates a file on disk and unless you know what to do, it won't be very useful to you. So maybe we could create specific postprocessing steps in bootstrap, for example to open a web browser with the gathered profile after running samply. But even that is not that easy, because what if you run more than one benchmark? :)

Doing profiling and benchmarking does require some specific knowledge at the moment. We could make it easier for the user, but I'm not sure if that code belongs here or into rustc-perf, and how far should we go in this.